### PR TITLE
refactor: implement schemas and all* methods in AsyncAPIDocument class

### PR DESCRIPTION
**Pull request by magicmatatjahu**
**Description**

Implement `schemas`, `allChannels`, `allServers`, `allOperations`, `allMessages`, `allSchemas` methods in the AsyncAPIDocument class.

cc @smoya 

**Related issue(s)**
Part of #481 
Part of #482

### CODE REVIEW

File: src/models/v2/asyncapi.ts [hunk 1]
```diff
@@ -76,7 +77,16 @@ export class AsyncAPIDocument extends BaseModel<v2.AsyncAPIObject> implements As
...
 
   schemas(): SchemasInterface {
-    return new Schemas([]);
+    const schemas: Set<SchemaInterface> = new Set();
+    function callback(schema: SchemaInterface) {
+      if (!schemas.has(schema.json())) {
+        schemas.add(schema);
+      }
+    }
+    // return only schemas used in channels ("active" schemas)
```
Comment (smoya) [inline src/models/v2/asyncapi.ts:86]: This is raising a hidden behaviour I didn't think when designing the API. Are we sure this is the behavior all users will need? Once i have the "active" schemas, how do I know the ones are not active?

I think we should rather return **all schemas** and add a filter in the `Schemas` collection such as `FilterByActive()` and `FilterByInactive` or `FilterByInUse()`/`FilterByNotInUse()` or whatever.

Comment (magicmatatjahu) [reply src/models/v2/asyncapi.ts:86]: If someone will need also inactive schemas can write:

```ts
    const schemas: Set<SchemaInterface> = new Set();
    function callback(schema: SchemaInterface) {
      if (!schemas.has(schema.json())) {
        schemas.add(schema);
      }
    }
    traverseAsyncApiDocument(this, callback);
    const schemas = new Schemas(Array.from(schemas));
```

Comment (smoya) [reply src/models/v2/asyncapi.ts:86]: Sure they can but I don't think we should take this decision on return only the ones used by channels but rather all of them. Then, thanks to filters, the could filter after all.

Comment (magicmatatjahu) [reply src/models/v2/asyncapi.ts:86]: 🤷🏼

Comment (smoya) [reply src/models/v2/asyncapi.ts:86]: We had a chat on Slack about it. As it ended up making sense, I'm gonna create those filters not only for `Schemas` but also for `Channels`, `Operations`, `Messages`, and `Servers`, also a PR to `parser-api` reflecting the change.
Can you @magicmatatjahu then make this method to return all schemas?

Comment (smoya) [reply src/models/v2/asyncapi.ts:86]: https://github.com/asyncapi/parser-js/pull/619

**Review by jonaslagoni — APPROVED** [review]

File: src/models/v2/asyncapi.ts (hunk 1 above)
Comment (smoya) [reply src/models/v2/asyncapi.ts:86]: @magicmatatjahu would you have some time to review the PR mentioned in my previous component, so we can unblock this? 🙏

Comment (smoya) [reply src/models/v2/asyncapi.ts:86]: I think you can adapt the code of this PR to consider https://github.com/asyncapi/parser-js/pull/619 is gonna happen so we can merge this right? It is not really blocked in fact.

Comment (magicmatatjahu) [reply src/models/v2/asyncapi.ts:86]: https://github.com/asyncapi/parser-js/pull/619#discussion_r985692933

File: src/models/v2/components.ts [hunk 2]
```diff
@@ -76,9 +76,7 @@ export class Components extends BaseModel<v2.ComponentsObject> implements Compon
   }
 
   operations(): OperationsInterface {
-    const operations: OperationInterface[] = [];
-    this.channels().forEach(channel => operations.push(...channel.operations().all()));
-    return new Operations(operations);
+    return new Operations([]);
```
Comment (magicmatatjahu) [inline src/models/v2/components.ts:79]: We don't have `operations` in `components` in v2, so we should return empty collection.

File: src/models/v2/asyncapi.ts [hunk 3]
```diff
@@ -91,7 +96,60 @@ export class AsyncAPIDocument extends BaseModel<v2.AsyncAPIObject> implements As
     return this.createModel(Components, this._json.components || {}, { pointer: '/components' });
   }
 
+  allServers(): ServersInterface {
```
[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply src/models/v2/asyncapi.ts:99]: In your example you will have two servers, `servers.myServer` and `components.servers.anotherServer`. We only compare the data of given object and we base on that filtering of these same "references".

[the remark this answers is not in the document]
Comment (smoya) [reply src/models/v2/asyncapi.ts:99]: So we compare the references and not the value of the json object?

[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply src/models/v2/asyncapi.ts:99]: yep, these same references == these same JSON objects

[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply src/models/v2/asyncapi.ts:99]: You can check tests for `allSchemas` or `allMessages` :)

[the remark this answers is not in the document]
Comment (smoya) [reply src/models/v2/asyncapi.ts:99]: Then, all good! Thanks for the explanation.

**Review by smoya — APPROVED** [review]
Awesome work @magicmatatjahu ! LGTM!

**Pull request merged**